lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,193 @@
+        path_excluding_numeric_indicies = data[:path].select { |p| p.is_a?(String) }
+
+        start_time = current_time
+        start_time_monotonic = current_time_monotonic
+
+        result = block.call
+        duration = current_time_monotonic - start_time_monotonic
+
+        data[:query].context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[context_key][path_excluding_numeric_indicies] ||= []
+          ns[context_key][path_excluding_numeric_indicies] << { start_time: start_time, duration: duration }
+        end
+
+        result
+      end
+
+      def runtime_metrics_enabled?(context)
+        return false unless context
+        !!context.namespace(CONTEXT_NAMESPACE)[RUNTIME_METRICS_ENABLED]
+      end
+    end
+
+    def initialize(query_or_multiplex)
+      super
+
+      query_or_multiplex.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+        ns[ANALYZER_INSTANCE_KEY] = self
+      end
+
+      @query = query_or_multiplex
+      @static_query_metrics = nil
+      @field_timing_callbacks = []
+    end
+
+    def analyze?
+      query.valid? && query.context[:skip_graphql_metrics_analysis] != true
+    end
+
+    def extract_query(runtime_metrics: {}, context:)
+      query_extracted(@static_query_metrics.merge(runtime_metrics))
+    end
+
+    # TODO: Apollo Tracing spec https://docs.google.com/document/d/1B0dR09CcN_M4yqezkJ7VFPI-mKgQIVCt9PiUcIMEVNI/edit
+    # Do so after integrating with Shopify/shopify, since we'll need another Instrumentation class to dig out values
+    # left in memory by Analyzer.
+
+    def on_enter_operation_definition(_node, _parent, visitor)
+      @static_query_metrics = {
+        operation_type: visitor.query.selected_operation.operation_type,
+        operation_name: visitor.query.selected_operation.name,
+      }
+    end
+
+    def on_leave_field(node, _parent, visitor)
+      static_metrics = {
+        field_name: node.name,
+        return_type_name: visitor.type_definition.name,
+        parent_type_name: visitor.parent_type_definition.name,
+        deprecated: visitor.field_definition.deprecation_reason.present?,
+        path: visitor.response_path,
+      }
+
+      if self.class.runtime_metrics_enabled?(visitor.query.context)
+        callback = -> (metrics) {
+          visitor.query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+            resolver_timings = ns[GRAPHQL_TRACING_FIELD_KEY][metrics[:path]]
+            lazy_field_timing = ns[GRAPHQL_TRACING_LAZY_FIELD_KEY][metrics[:path]]
+
+            metrics = metrics.merge(
+              resolver_timings: resolver_timings,
+              lazy_resolver_timings: lazy_field_timing,
+            )
+
+            field_extracted(metrics)
+          end
+        }
+
+        @field_timing_callbacks << FieldTimingCallback.new(static_metrics: static_metrics, callback: callback)
+      else
+        field_extracted(static_metrics)
+      end
+    end
+
+    def run_field_timing_callbacks
```

Comment:
gotta admit I don't fully grok the whole callback setup you've got going on here

Reply:
tldr, when used as instrumentation, an analyzer and tracing, the order of execution is:

* before_query (context setup)
* initialize (bit more context setup, instance vars setup)
  ^ I'm reconsidering where and how the setup logic is all done
* result (static analyzer results ready, including the opportunity to call extract_query for static query metrics if the analyzer is not being used as instrumentation)
* after_query (call query and field callbacks, now that we have all static and runtime metrics gathered)
* extract_query
* run_field_timing_callbacks
* calling field timing callback (n times)

When used as a simple analyzer, which doesn't gather or emit any runtime metrics (timings, arg values):
* initialize
* result
* extract_query